### DISCUSSION TITLE
Devcontainer support

### DISCUSSION THREAD

**Discussion by auser** [zed-industries/zed#9023] [Feature Requests]
My organization uses devcontainers extensively. I personally love using zed.

Are there plans on integrating devcontainer work with zed?

**Comment by Moshyfawn** [comment]
There's already a feature request for this in #5347.

**Reply by KarstenB** [reply]
The proper issue for DevContainer is actually #11473. Remote Development is a rather different use case.

**Reply by NewtonChutney** [reply]
@auser can you mark the above reply as the answer.. the discussions should move into the issue

**Comment by sikmir** [comment]
zed tasks + devcontainer cli works fine for me.

**Reply by kodmigor** [reply]
can you tell, please, how did you setup it?

**Reply by OlegChuev** [reply]
@kodmigor I believe you can use the following approach

1) Install devcontainers cli
```bash
npm install -g @devcontainers/cli
```

2) Create a `.devcontainer/devcontainer.json` file in your project root.

3) Create a `.zed/tasks.json` file in your project root with these tasks:

```json
[
  {
    "label": "DevContainer: Up",
    "command": "devcontainer up --workspace-folder ${ZED_WORKTREE_ROOT}",
    "use_new_terminal": true,
    "allow_concurrent_runs": false,
    "reveal": "always"
  },
  {
    "label": "DevContainer: Build",
    "command": "devcontainer build --workspace-folder ${ZED_WORKTREE_ROOT}",
    "use_new_terminal": true,
    "allow_concurrent_runs": false,
    "reveal": "always"
  },
  {
    "label": "DevContainer: Exec",
    "command": "devcontainer exec --workspace-folder ${ZED_WORKTREE_ROOT} ${1}",
    "use_new_terminal": true,
    "allow_concurrent_runs": true,
    "reveal": "always"
  },
  {
    "label": "DevContainer: Run User Commands",
    "command": "devcontainer run-user-commands --workspace-folder ${ZED_WORKTREE_ROOT}",
    "use_new_terminal": true,
    "allow_concurrent_runs": false,
    "reveal": "always"
  }
]
```


Usage:

1. Open your project in Zed
2. Press `Cmd/Ctrl + Shift + P` to open the command palette
3. Type `task: spawn` and select it
4. Choose “DevContainer: Build” to build your container
5. Use “DevContainer: Up” to start it
6. Run commands inside your container with “DevContainer: Exec”

Footprints: https://alessandro-annini.medium.com/lightning-fast-development-with-zed-and-dev-containers-11df1135b635

**Reply by kodmigor** [reply]
thanks! i'll try.

**Reply by WithHolm** [reply]
if you are on windows, the command might fail as double dash might gets "fixed" when running the command parameter making the command be `devcontainer build - workspace-folder ....`. if it happens, move the `build --workspace-folder ...` to args instead:
``` json
    "command": "devcontainer",
    "args": ["build --workspace-folder ${ZED_WORKTREE_ROOT}"],
```

**Reply by mateka** [reply]
Does code completion works? I would assume, that dependencies would be installed in devcontainer.

**Reply by twright** [reply]
> Does code completion works? I would assume, that dependencies would be installed in devcontainer.

It does not. Tasks only address launching containers (which you can do with plain docker compose), but not running the LSP in the container, which is the core functionality of devcontainers.

The best current workaround is to add a new service to the devcontainer running sshd, and connect to this using remote editing support, but proper devcontainer support would be much better.

**Reply by airtonix** [reply]
> @kodmigor I believe you can use the following approach

A core feature of devcontainers is that your editor divide it self into a client and a server, run the server component in your devcontainer.

**Comment by CoolGoose** [comment]
Any updates on this ? :)

**Reply by NewtonChutney** [reply]
follow here: https://github.com/zed-industries/zed/issues/11473
